src/target/riscv/batch.c

```diff
@@ -18,7 +18,13 @@
 
```

Comment:
My recommendation would be to implement the reset of the DMI scan delay in the batch via simpler, more straightforward code:

- Since there is currently at most one reset of `idle_count` in a given batch, I'd recommend to remove the the list, and instead use just a plain counter do determine at which point in the batch to reset .
- If you prefer to still have the possibility to specify different idle count for each scan in the batch, please allocate an integer value directly for each scan instead of a list: `out->idle_counts = malloc(sizeof(unsigned) * scans);`. The extra bit of used memory is, in my opinion, very good price to pay for code simplicity.
- Please, decrement the `r->reset_delays_wait` directly in the loop in `jtag_fill_queue()` as the scans are added into the jtag queue; instead of doing it in another module in function `maybe_add_delays_reset()`.

Reply:
@en-sc Thank you for explaining your changes even more.

In another words, you are proposing to create sub-batches within a batch in order to be able to manipulate idle counts for a subset of the scans in the batch at once. 

I am sorry but I must respectfully push back on this change.

This would bring a new additional level of complexity into the batch processing. Considering that we do not have an easy way how to test this part of OpenOCD's due to Spike's limitations (as discussed above [1](https://github.com/riscv-collab/riscv-openocd/pull/1028#issuecomment-2046976554), [2](https://github.com/riscv-collab/riscv-openocd/pull/1028#issuecomment-2049110776)), I would like to absolutely avoid making this part of OpenOCD more complex when the same functionality can be delivered in a simpler, more direct way, without creating sub-structures inside the batches. Please see my proposal below.

Also the fact that long explanation was needed, IMO, underlines the logical, principal complexity of the proposed change.

Proposal of a simpler solution:

- For each scan in a batch, store the type of the scan: 
  - The types would be:
    - Simple DMI read or write(analogous to `dmi_read()` or `dmi_write()`) 
    - DMI read or write that triggers abstract command execution (analogous to `dmi_read_exec()` or `dmi_write_exec()`)
    - DMI read or write that triggers a system bus transaction
- Each scan uses the idle count that is valid at the time of starting of the batch
- `riscv_batch_run()` would read the current idle values directly from `r->dmi_busy_delay`, `r->bus_master_write_delay`, etc. That is, single source of authority.
- Re-execution of the batch is possible - with up to date idle counts
- Parameter `idle_count` in `riscv_batch_alloc()` would be removed - no longer needed.
- Reset of learned idle values would be handled by decrementing `r->reset_delays_wait` in fill_jtag_queue().

This proposal covers all the use cases discussed above:

- Different idle values for different scans in the batch
- Possibility to re-execute an existing batch batch - with the same or different idle values
- Reset of learned delays during batch execution

Please, let me know what you think.

Reply:
@JanMatCodasip, thank you for the suggestion!

My proposal is more complicated than the one you suggest. I would like to try to explain my motivation for this.

> This would bring a new additional level of complexity into the batch processing.

This is true, however, I believe the changes to the processing itself (running the batch) are minimal (just iterate over list entries before iterating over batch elements).

> Considering that we do not have an easy way how to test this part of OpenOCD's due to Spike's limitations (as discussed above [1](https://github.com/riscv-collab/riscv-openocd/pull/1028#issuecomment-2046976554), [2](https://github.com/riscv-collab/riscv-openocd/pull/1028#issuecomment-2049110776)).

This is precisely why I have posted this patch first. With the ability to reset delays on any scan in the batch tests from `riscv-tests/debug` do test this code.

> Also the fact that long explanation was needed, IMO, underlines the principal complexity of the proposed change.

I agree, the implementation is somewhat complex, but the interface is simple. AFAIU, `riscv_batch_change_idle_used_from_scan()` is the one function that could have been simpler. Though it is quite short.

Regarding your proposal:

* I would like for the batch to be decoupled from target. IMHO, the reference to `struct target` should be replaced by a reference to `struct dtm *` (to be implemented). This is more straightforward, since the batches access DM, not the target itself. Reference to target will complicated accessing DM registers via a batch (which I believe should be the only access method) when switching from one target to another or when the target is not yet examined.
* I don't think the set of delay types you propose will be enough. What about Quick Access? What if a target is expected to change state in between the operations (e.g. resuming)?

**TLDR**: I believe the interface you propose will prove to narrow to be used in all the cases, so it will not be possible to use throughout the codebase.

However, I'm obviously biased and, if you insist, I will implement your suggestion instead.

Reply:
@en-sc Thank you for taking time to explain your ideas further.

> I don't think the set of delay types you propose will be enough. What about Quick Access? What if a target is expected to change state in between the operations (e.g. resuming)?

Agreed. The set of delays in my comment is just an example. For each type of operation that can take a different amount of time (sysbus read, sysbus write, quick access, ...), OpenOCD should keep a separate delay value.

> I would like for the batch to be decoupled from target. IMHO, the reference to struct target should be replaced by a reference to struct dtm * (to be implemented). This is more straightforward, since the batches access DM, not the target itself.

I am neutral to decoupling the target and the batch (I am not against it and at the same time not too eager). 

Below is my updated proposal that can work when batches are decoupled from `target`: The currently valid delays can be passed to riscv_batch_run():

```c
struct riscv_learned_delays
{
	unsigned int dmi_busy_delay;
	unsigned int ac_busy_delay;
	/* ... add other delays: system bus, quick access ... */

	/* when decremented from 1 to 0, the learned delays above are cleared to 0 */
	int reset_delays_wait;
};

struct riscv_info {
	/* ... */
	struct riscv_learned_delays delays;
	/* ... */
};	

// fill batch:
riscv_batch_add_dm_write(batch, DM_SBCS, sbcs_value, DELAY_BASIC_DMI);
riscv_batch_add_dm_write(batch, DM_SBADDRESS0, addr, DELAY_SYSBUS_READ);
// ...

// execute batch:
riscv_batch_run(batch, /* struct riscv_learned_delays* */ &delays);

```

> ... I agree, the implementation is somewhat complex, but the interface is simple.  ...

I am sorry to disagree with this statement. The batch interface would become more difficult to use: the batch user would need to keep track of sub-groups in the batch so that the delays can be properly updated when re-executing the batch.

Furthermore, the subgroups in the batch do not scale well if different types of delays are mixed inside the batch. Imagine a developer wishes to implement writing to memory using progbufsize=1 via a batch (a real-word scenario, https://github.com/riscv-collab/riscv-openocd/issues/333). The resulting batch would look like this in RV32:

```
write to DATA0 (short delay)
write to COMMAND (write address to a GPR, medium delay)
write to DATA0 (short delay)
write to COMMAND (write data to a GPR + execute program buffer, long delay)
<repeat n times>
```

Due to different delays for each of the command, there would be too many sub-groups. The caller would need to go through all the sub-groups and update the delays before re-executing the batch.

Logically, the delay for a DMI operation depends on the nature of the operation itself: on what types of operations get triggered by the DMI scan. That's why I strongly believe that the delays should be assigned to each operation, and subgroups go against this principle (demonstrated on the example with progbufsize=1 above).

> I believe the interface you propose will prove to narrow to be used in all the cases, so it will not be possible to use throughout the codebase.

Please, let me know what concrete use cases have not yet been addressed, so that we can together find a way how to address them using simple, straightforward code. Thank you.

---
---

Off-topic, unrelated to this PR:

> ... via a batch (which I believe should be the only access method) ...

[Off-topic, unrelated to this PR:] Do I understand it correctly that your future aim would be to use batch access for all DMI accesses? If so, then that sounds a bit too radical to me. In my opinion, both the APIs should remain in place since each of them supports different use cases:

- `dmi_read()`, `dmi_write()`, `...`: For algorithms where performance is not crucial, but where robust execution of DMI operations is desired. DMI busy states are handled gracefully without the need to care about it in the user code. Therefore, the user algorithms can remain very simple (easier to read and maintain, less bug-prone, ...).
- `riscv_batch_*`: For algorithms where performance is important. The user code must be ready for handling errors occurring in the middle of the batch, and know how to properly recover from them. Therefore, the user-written algorithms using batches will generally lead to more complex code, which is a tradeoff for the improved performance.

If you'd like to make larger changes to these two fundamental APIs at some future point, please, it would be good to discuss it first.

Reply:
> Off-topic, unrelated to this PR:
> 
> > ... via a batch (which I believe should be the only access method) ...
> 
> Do I understand it correctly that your future aim would be to use batch access for all DMI accesses? If so, then that sounds a bit too radical to me. In my opinion, both the APIs should remain in place since each of them supports different use cases:
> 
> * `dmi_read()`, `dmi_write()`, `...`: For algorithms where performance is not crucial, but where robust execution of DMI operations is desired. DMI busy states are handled gracefully without the need to care about it in the user code. Therefore, the user algorithms can remain very simple (easier to read and maintain, less bug-prone, ...).
> * `riscv_batch_*`: For algorithms where performance is important. The user code must be ready for handling errors occurring in the middle of the batch, and know how to properly recover from them. Therefore, the user-written algorithms using batches will generally lead to more complex code, which is a tradeoff for the improved performance.
> 
> If you'd like to make larger changes to these two fundamental APIs at some future point, please, it would be good to discuss it first.

The thing is, `dmi_read()`, `dmi_write()`, `...` utilize VLA to store fields. This is against the guidelines (from `doc/manual/style.txt`)
```
- use malloc() to create dynamic arrays. Do @b not use @c alloca
or variable length arrays on the stack. non-MMU hosts(uClinux) and
pthreads require modest and predictable stack usage.
```
So these should be eventually replaced by a `malloc()/calloc()`. And after this change is done it would be just easier to use appropriate batch of two accesses in `dmi_read()`, `dmi_write()`, `...`. (no change to API).

Reply:
I have simplified the commit. I will implement scan types as you suggested in subsequent commits. Please, take a look.

Reply:
> I have simplified the commit. I will implement scan types as you suggested in subsequent commits. Please, take a look.

Thank you. I have posted just few minor coding suggestions below.

> The thing is, dmi_read(), dmi_write(), ... utilize VLA to store fields. (...) 
> (...) And after this change is done it would be just easier to use appropriate batch of two accesses in dmi_read(), dmi_write(), .... (no change to API).

Thank you for explanation. I have not yet studied this part of the code in detail (especially the use of VLAs). So I do not yet have clear view whether it would be beneficial to re-implement dmi_read(), dmi_write() et. al. on top of batches, or whether to choose a different approach.